### ISSUE TITLE
History traversal - improve spec compliance

### ISSUE THREAD

**Issue by gterzian** [servo/servo#23479]
@asajeffrey @cbrewster 

Related: https://github.com/servo/servo/issues/10994

I happened to be looking into setting the [request-history-navigation-flag](https://fetch.spec.whatwg.org/#concept-request-history-navigation-flag) as part of https://github.com/servo/servo/pull/23368, and ended-up looking once more into the structure of the current implementation history traversal algorithm.

It appears to me that most of the "logic" happens in the constellation, with a few back and forth with script with things like `ConstellationControlMsg::UpdateHistoryState`. 

Also, it appears to me that the actual "navigation" happens straight in the constellation, where pipelines are updated and I think also potentially reloaded. 

Finally, it appears to me that we're not doing a bunch of stuff that would have to happen in script, such as firing `popstate` or `hashchange` events. 

From a high-level perspective, I think the main problem with the current implementation is that we're not doing Step 5 of [traverse-the-history-by-a-delta](https://html.spec.whatwg.org/multipage/#traverse-the-history-by-a-delta) which reads:

> Queue a task that consists of running the following substeps. The relevant event loop is that of the specified browsing context's active document. The task source for the queued task is the history traversal task source.

And those substeps involve:
1. Potentially cancelling any ongoing navigation of the "specified browsing context"
2. Potentially unloading the currently active document of the "specified browsing context"
3. And finally, running all the [traverse-the-history](https://html.spec.whatwg.org/multipage/#traverse-the-history) steps. 

Step 3 includes stuff like "navigating" to perform an "entry update" in case the entry doesn't hold a document anymore. It also involves stuff like update the state of the history, firing events, and so on. 

The nice thing about doing all of these steps as a task on the script-thread, is that it means things like "reloading" actually go through the normal navigation flow as it originates inside of script, as opposed to a custom algorithm implemented fully inside the constellation. 

This matters for things like setting a " history-navigation flag" to the request, since now you'd have to, I think, integrate it "by hand" into the algorithm in constellation, whereas otherwise it could fit in more easily into a single unified workflow that always starts inside the script-thread.

Let me know what you think, I just had a quick look around so I probably missed a few things...

**Comment by gterzian** [comment]
Also related is the concept of the [session-history-traversal-queue](https://html.spec.whatwg.org/multipage/history.html#session-history-traversal-queue) and the [session-history-event-loop](https://html.spec.whatwg.org/multipage/history.html#session-history-event-loop)

> Each top-level browsing context has a session history traversal queue, initially empty, to which tasks can be added.

> Each top-level browsing context, when created, must begin running the following algorithm, known as the session history event loop for that top-level browsing context, in parallel

So since it's "in-parallel", one could say the constellation is a good place to do it, and in some way the current implementation is essentially this "session-history-event-loop" running in the constellation. 

> To traverse the history by a delta delta, the user agent must append a task to this top-level browsing context's session history traversal queue, the task consisting of running the following steps

Which is really what we do here, since w send a message from script to the constellation:
https://github.com/servo/servo/blob/0a9a222356ae5b491885edd381fb5b821074ee1c/components/script/dom/history.rs#L74

However what's missing is the "queue a task back on an event-loop, not necessarily the one from which the original message above originated, and run the rest of the traversal algorithm". 

It might very well be that the various parts of the current implementation, including the back and forth with various messages between constellation and script, is actually fully spec compliant. However I'm pretty sure it would be more robust the follow the spec for closely and actually queue a task back on an event-loop, and take it from there. So it essentially puts most of the logic to run back on an event-loop, with messages to the constellation for some things like actually loading an url.

**Comment by asajeffrey** [comment]
Yeah, the current implementation is spread out a lot, and involves more back-and-forth between script and the constellation. Some of this is inevitable, as there are four threads involved: the constellation, the script thread that performed the navigation, the script thread being navigated away from, and the script thread being navigated to. Not to mention that the page being navigated to might have been evicted, so navigating to it may involve creating script threads and reloading the document,

Part of the reason for leaving this alone was to give the spec some time to settle down, there are a lot of open issues around navigagtion at https://github.com/whatwg/html/labels/topic%3A%20navigation. This may be cowardice on my part though.

**Comment by cbrewster** [comment]
I would also like to add that its really difficult to know what to do in a script thread when a history traversal is requested because history traversal always operates on the joint session history. A single script thread cannot really reason about the joint session history as that may require knowing info from other script threads. Naturally the constellation is a place that has all the info that we need.

I'm not entirely convinced that the current spec really reflects the reality of session history that is implemented in The Browsers, but it's been a while since I have read the history traversal spec so it may have improved since then.

**Comment by gterzian** [comment]
@cbrewster 
> I would also like to add that its really difficult to know what to do in a script thread when a history traversal is requested because history traversal always operates on the joint session history. A single script thread cannot really reason about the joint session history as that may require knowing info from other script threads. Naturally the constellation is a place that has all the info that we need.

Yes and that is the role of what the spec refers to as the [session-history-event-loop](https://html.spec.whatwg.org/multipage/history.html#session-history-event-loop), which is supposed to run "in-parallel" of the script-event-loop.

So I would say running what is the equivalent of this "session-history-event-loop" in the constellation not only makes sense from your experience in implementing it in Servo, it's also what the spec points towards.

See
> The session history event loop helps coordinate cross-browsing-context transitions of the joint session history: since each browsing context might, at any particular time, have a different event loop (this can happen if the user navigates from example.com to shop.example), transitions would otherwise have to involve cross-event-loop synchronization.


@asajeffrey 

> there are four threads involved: the constellation, the script thread that performed the navigation, the script thread being navigated away from, and the script thread being navigated to. Not to mention that the page being navigated to might have been evicted, so navigating to it may involve creating script threads and reloading the document

Yes and the spec does take this into account, for example Step 1 of [traverse-the-history](https://html.spec.whatwg.org/multipage/browsing-the-web.html#traverse-the-history) that reads:

> If entry no longer holds a Document object, then (...) Navigate the browsing context to request to perform an entry update of entry.

I would say the main difference between the spec and the current implementation, and that which we could at some point address, is that Step 5 of [traverse-the-history-by-a-delta)](https://html.spec.whatwg.org/multipage/history.html#traverse-the-history-by-a-delta) basically moves the algorithm from "in-parallel"(read: the constellation) steps related to manipulating the "joint-session-history" to continuing on the event-loop of the "active document" of the "specified browsing context" by way of queuing a task. 

A specific example:

Currently constellation sends a message to script to do the following:

https://github.com/servo/servo/blob/a5506365d6f1e8a75db7e7fa2f9e0b1ef73972ee/components/script/script_thread.rs#L2289

I'm assuming this corresponds to Step 3 of "traverse the history" that reads "If appropriate, update the current entry in the browsing context's Document object's History object to reflect any state that the user agent wishes to persist."

Currently the code is handling the "cannot find that document/window" case with a `"update history state after pipeline {} closed."`

So this case should simply never happen if the entire "traverse the history" was done inside a single task, since if the "specified entry" doesn't hold a document anymore, the spec hits the "Navigate the browsing context to request to perform an entry update of entry" branch, and as the spec notes:

> The "navigate" algorithm reinvokes this "traverse" algorithm to complete the traversal, at which point entry holds a Document object.

Therefore we should never hit a `"update history state after pipeline {} closed."` point, and I think queuing a task on an event-loop to run the  "traverse the history"(not "traverse-the-history-by-a-delta", which starts in parallel and then should go to an event-loop to perform "traverse the history") should make these issues go away(since the entire algorithm would be serialized onto a script-thread where appropriate). 

----

Finally it might also be that what we're hitting in Servo is the "we've closed the script-thread on which that particular entry in the joint-session-history was running", which would be different from "the entry doesn't hold a document anymore and we need to navigate". It might be then that we're too aggressive in closing pipelines/script-thread in relation to their use in the joint-session-history... 

In any case, in my experience, when you realize the spec seems to imply a certain structure that doesn't seem to fit what you currently have, when you make changes in the direction of what the spec suggests, it usually solves a bunch of issues that previously seemed like weird intermittent bugs...

**Comment by asajeffrey** [comment]
Hmm, I think the reason why there's a disconnect between what Servo does and the spec is that there's some tasks that the spec says should be performed on the event loop of the active document of the specified browsing context but which Servo needs to be handled by the constellation.

For example, step 5.3 of https://html.spec.whatwg.org/multipage/history.html#traverse-the-history-by-a-delta is "Traverse the history of the specified browsing context to the specified entry with the history-navigation flag set," and step 4.3 of https://html.spec.whatwg.org/multipage/browsing-the-web.html#traverse-the-history is "Set the active document of the browsing context to entry's Document object."

Unfortunately, in Servo, setting the active document isn't something that a script thread can do directly, it has to message the constellation, which updates its state and then messages all the affected pipelines.

That said, the spec is now a lot closer to what Servo does than it used to be, and reorganizing the code to more closely match the spec is A Good Thing, as long as the spec is reasonably stable.

@annevk do you know if there's any planned major changes to this bit of navigation any time soon?

**Comment by gterzian** [comment]
> Unfortunately, in Servo, setting the active document isn't something that a script thread can do directly, it has to message the constellation, which updates its state and then messages all the affected pipelines.

Yes, and  what I'm thinking of is to move the "center of gravity" of our implementation of the algorithm to a script-thread via a task queued using the "history-traversal-task-source", and message the constellation from that script-thread when necessary. 

The main reason would be that I think it would matter less to have the constellation running and handling other messages while the "history traversal(the last part that is speced to run inside a task)" runs on a script-thread, than to have the script-thread continue to run while the main logic is taking place in the constellation with several messages used to affect the world of the script-thread. 

For example in [traverse-the-history](https://html.spec.whatwg.org/multipage/browsing-the-web.html#traverse-the-history), we have:

-  Step 4.1 "Remove any tasks queued by the history traversal task source that are associated with any Document objects in the top-level browsing context's document family.",
- 4.6.4 "Fire an event named pageshow at the Document object's relevant global object"
- 5 Set the document's URL to entry's URL
- 13 Set history.state to state.
- 16 "If the non-blocking events flag is not set, then run the following substeps immediately. Otherwise, the non-blocking events flag flag is set; queue a task to run the following substeps instead."
- 16.1 "If state changed is true, then fire an event named popstate at the Document "

And those are all examples of things affecting the document, which are specced to all happen in the same task queued using the "history-traversal-task-source". When those steps mention "queuing a task", that's that "Dom manipulation task-source". 

So I think that if one tries to implement all those steps from the constellation, using messaging to the relevant script-thread, you risk a lot of intermittent issues simply because script-thread will keep running in the meantime, and also the various "queue a subsequent task in some cases, in others fire an event immediately" are hard to express via messaging from the constellation, since that's a bit like queuing a task in itself. Finally, the "Remove any tasks queued by the history traversal task source" step would seem to really require going down the task-source route, since otherwise you'd have to do something manually by ignoring some messages from the constellation in some cases. 

On the other hand, we also have things like:

- 4.3 "Set the active document of the browsing context to entry's Document object."(mentioned earlier)
- 4.4 "Clear any browsing context names of all entries in the history that are associated with Document objects with the same origin as the new active document and that are contiguous with entry."
- 9 "Set the current entry to entry."
- 15 "Set entry's Document object's latest entry to entry."

Which seem like examples of stuff requiring to happen in the constellation, and I'd argue there should be no problem sending either one, or several, messages from the relevant script-thread to the constellation, with the constellation continuously running in the meantime.

Finally there steps like:

- 6 "If entry has a URL whose fragment differs from that of the current entry's when compared in a case-sensitive manner, and the two share the same Document object, then let hash changed be true, and let old URL be the current entry's URL and new URL be entry's URL. Otherwise, let hash changed be false."

That would seem fit to happen on the constellation, before the actually "history-traversal-task_source" is queued on a relevant script-thread, while passing the results of such operations along to the script-thread.

Having said all that, it's usually when you try to implement these workflows that you run into the real issues :)

**Comment by annevk** [comment]
@asajeffrey there's https://github.com/whatwg/html/pull/4664 and there are some more plans to enable `Cross-Origin-Opener-Policy` and friends, but not sure on timeline.

**Comment by asajeffrey** [comment]
@annevk okay, sounds like it's stable enough for us to redo the current implementation, thanks!

**Comment by annevk** [comment]
Yeah, though please have a look through the open issues around navigation and history so you don't do wasted work.

**Comment by asajeffrey** [comment]
@gterzian yes, that sounds right. We could at least make the implementation follow the latest spec a bit more. The main thing to look out for is deadlock: https://github.com/servo/servo/blob/ced2668c6c4172069ba4789b805137d7c1375ce3/components/constellation/constellation.rs#L64-L90

We can have script block on the constellation but not the other way round.
